File: app/src/main/java/com/dmi/perfectreader/layout/wordbreak/TeXHyphenator.java
package com.dmi.perfectreader.layout.wordbreak;

import com.carrotsearch.hppc.ByteArrayList;
import com.carrotsearch.hppc.CharArrayList;
import com.dmi.util.annotation.Reusable;
import com.dmi.util.cache.ReuseCache;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.List;

import static com.dmi.util.cache.ReuseCache.reuser;
import static com.google.common.io.CharStreams.readLines;
import static java.lang.Math.min;
import static java.util.Arrays.copyOf;

public class TeXHyphenator {
    /**
     * алгоритм: https://habrahabr.ru/post/138088/
     */

    private static final char EDGE_OF_WORD = '.';

    private final Patterns patterns;
    private final CharsLevels exceptions;

    private TeXHyphenator(Patterns patterns, CharsLevels exceptions) {
        this.patterns = patterns;
        this.exceptions = exceptions;
    }

    public WordBreaker.WordBreaks breakWord(CharSequence text, int beginIndex, int endIndex) {
        int length = endIndex - beginIndex;
        PatternLevels wordLevels = Reusables.wordLevels(length);

        if (length >= 3 && (patterns.count > 0 || exceptions.size() > 0)) {
            PatternChars chars = Reusables.wordChars();
            if (!applyException(text, beginIndex, endIndex, wordLevels, chars)) {
                applyPatterns(text, beginIndex, endIndex, wordLevels, chars);
            }
        }

        return Reusables.wordBreaks(wordLevels, beginIndex);
    }

    private boolean applyException(CharSequence text, int beginIndex, int endIndex, PatternLevels wordLevels, PatternChars chars) {
        chars.reset(text, beginIndex, endIndex, true, true);
        PatternLevels levels = exceptions.get(chars);
        if (levels != null) {
            levels.applyTo(wordLevels, 0);
            return true;
        } else {
            return false;
        }
    }

    private void applyPatterns(CharSequence text, int beginIndex, int endIndex, PatternLevels wordLevels, PatternChars chars) {
        for (int begin = beginIndex; begin < endIndex; begin++) {
            for (int end = min(endIndex, begin + patterns.maxLength()); end > begin; end--) {
                if (begin == beginIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, true, false);
                }
                if (end == endIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, false, true);
                }
                if (begin == beginIndex && end == endIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, true, true);
                }
                applyLevels(text, beginIndex, wordLevels, chars, begin, end, false, false);
            }
        }
    }

    private void applyLevels(CharSequence text, int beginIndex, PatternLevels wordLevels, PatternChars chars, int begin, int end,
                             boolean atWordBegin, boolean atWordEnd
    ) {
        chars.reset(text, begin, end, atWordBegin, atWordEnd);
        PatternLevels levels = patterns.levelsFor(chars);
        if (levels != null) {
            levels.applyTo(wordLevels, begin - beginIndex);
        }
    }

    public static class Builder {
        private Patterns patterns = new Patterns();
        private CharsLevels exceptions = new CharsLevels();

        public Builder addPatternsFrom(InputStream is) throws IOException {
            List<String> patterns = readLines(new InputStreamReader(is));
            for (String pattern : patterns) {
                addPattern(pattern);
            }
            return this;
        }

        public Builder addExceptionsFrom(InputStream is) throws IOException {
            List<String> patterns = readLines(new InputStreamReader(is));
            for (String pattern : patterns) {
                addException(pattern);
            }
            return this;
        }

        public Builder addPattern(String pattern) {
            CharArrayList letters = new CharArrayList();
            PatternLevels levels = new PatternLevels();
            levels.resize(pattern.length() + 1);

            boolean atWordBegin = false;
            boolean atWordEnd = false;

            for (int i = 0; i < pattern.length(); i++) {
                char ch = pattern.charAt(i);
                if (ch == EDGE_OF_WORD) {
                    if (i == 0) {
                        atWordBegin = true;
                    } else if (i == pattern.length() - 1) {
                        atWordEnd = true;
                    }
                } else if ('0' <= ch && ch <= '9') {
                    levels.set(letters.size(), (byte) (ch - '0'));
                } else {
                    letters.add(ch);
                }
            }

            String patternStr = new String(letters.buffer, 0, letters.elementsCount);
            PatternChars patternChars = new PatternChars(patternStr, 0, patternStr.length(), atWordBegin, atWordEnd);

            levels.resize(letters.size() + 1);
            levels.trimToSize();

            patterns.put(patternChars, levels);

            return this;
        }

        public Builder addException(String exception) {
            CharArrayList letters = new CharArrayList();
            PatternLevels levels = new PatternLevels();
            levels.resize(exception.length() + 1);

            for (int i = 0; i < exception.length(); i++) {
                char ch = exception.charAt(i);
                if (ch == '-') {
                    levels.set(letters.size(), (byte) 9);
                } else {
                    letters.add(ch);
                }
            }

            String patternStr = new String(letters.buffer, 0, letters.elementsCount);
            PatternChars patternChars = new PatternChars(patternStr, 0, patternStr.length(), true, true);

            levels.resize(letters.size() + 1);
            levels.trimToSize();

            exceptions.put(patternChars, levels);

            return this;
        }

        public TeXHyphenator build() {
            return new TeXHyphenator(patterns, exceptions);
        }
    }

    private static class Patterns {
        private CharsLevels[] lengthToCharsLevels = new CharsLevels[16];
        private int maxLength = 0;
        public int count = 0;

        public void put(PatternChars chars, PatternLevels levels) {
            int length = chars.length;
            if (length > lengthToCharsLevels.length) {
                lengthToCharsLevels = copyOf(lengthToCharsLevels, length >> 1);
            }
            if (length > maxLength) {
                maxLength = length;
            }

            CharsLevels charsLevels = lengthToCharsLevels[length];
            if (charsLevels == null) {
                charsLevels = new CharsLevels();
                lengthToCharsLevels[length] = charsLevels;
            }

            charsLevels.put(chars, levels);
            count++;
        }

        public int maxLength() {
            return maxLength;
        }

        public PatternLevels levelsFor(PatternChars chars) {
            CharsLevels charsLevels = lengthToCharsLevels[chars.length];
            return charsLevels != null ? charsLevels.get(chars) : null;
        }
    }

    private static class CharsLevels extends HashMap<PatternChars, PatternLevels> {}

    @Reusable
    private static class PatternChars {
        private CharSequence str;
        private int begin;
        private int end;
        private boolean atWordBegin;
        private boolean atWordEnd;

        public int length;
        private int hashCode;

        public PatternChars() {
        }

        public PatternChars(CharSequence str, int begin, int end, boolean atWordBegin, boolean atWordEnd) {
            reset(str, begin, end, atWordBegin, atWordEnd);
        }

        public void reset(CharSequence str, int begin, int end, boolean atWordBegin, boolean atWordEnd) {
            this.str = str;
            this.begin = begin;
            this.end = end;
            this.atWordBegin = atWordBegin;
            this.atWordEnd = atWordEnd;
            this.length = end - begin;

            computeHashCode();
        }

        @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
        @Override
        public boolean equals(Object o) {
            PatternChars other = (PatternChars) o;

            if (other.length != length || other.atWordBegin != atWordBegin || other.atWordEnd != atWordEnd) {
                return false;
            }

            for (int i = begin, j = other.begin; i < end; i++, j++) {
                if (str.charAt(i) != other.str.charAt(j)) {
                    return false;
                }
            }

            return true;
        }

        @Override
        public int hashCode() {
            return hashCode;
        }

        private void computeHashCode() {
            hashCode = 0;
            for (int i = begin; i < end; ++i) {
                hashCode = 31 * hashCode + str.charAt(i);
            }
            hashCode = 31 * hashCode + (atWordBegin ? 1 : 0);
            hashCode = 31 * hashCode + (atWordEnd ? 1 : 0);
        }
    }

    @Reusable
    private static class PatternLevels extends ByteArrayList {
        void applyTo(PatternLevels levels, int beginIndex) {
            for (int i = 0, j = beginIndex; i < elementsCount && j < levels.elementsCount; i++, j++) {
                byte level = get(i);
                if (level > levels.get(j)) {
                    levels.set(j, level);
                }
            }
        }
    }

    @Reusable
    private static class WordBreaksImpl implements WordBreaker.WordBreaks {
        private PatternLevels levels;
        private int beginIndex;

        public void reset(PatternLevels levels, int beginIndex) {
            this.levels = levels;
            this.beginIndex = beginIndex;
        }

        @Override
        public boolean canBreakBefore(int index) {
            int wordIndex = index - beginIndex;
            // переносить одну букву нельзя
            boolean isMiddleBreak = wordIndex >= 2 && wordIndex < levels.elementsCount - 2;
            return isMiddleBreak && levels.get(wordIndex) % 2 != 0;
        }
    }

    private static class Reusables {
        private static final ReuseCache.Reuser<PatternChars> wordChars = reuser(PatternChars::new);
        private static final ReuseCache.Reuser<PatternLevels> wordLevels = reuser(PatternLevels::new);
        private static final ReuseCache.Reuser<WordBreaksImpl> wordBreaks = reuser(WordBreaksImpl::new);

        public static PatternChars wordChars() {
            return wordChars.reuse();
        }

        public static PatternLevels wordLevels(int wordLength) {
            PatternLevels value = wordLevels.reuse();
            value.clear();
            value.resize(wordLength + 1);
            return value;
        }

        public static WordBreaker.WordBreaks wordBreaks(PatternLevels levels, int beginIndex) {
            WordBreaksImpl value = wordBreaks.reuse();
            value.reset(levels, beginIndex);
            return value;
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/layout/wordbreak/TeXHyphenator.kt
package com.dmi.perfectreader.layout.wordbreak

import com.carrotsearch.hppc.ByteArrayList
import com.carrotsearch.hppc.CharArrayList
import com.dmi.util.annotation.Reusable
import com.dmi.util.cache.ReuseCache.reuser
import com.google.common.base.Charsets
import com.google.common.io.CharStreams.readLines
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.lang.Math.min
import java.util.*
import java.util.Arrays.copyOf

class TeXHyphenator private constructor(private val patterns: TeXHyphenator.Patterns, private val exceptions: TeXHyphenator.CharsLevels) {
    fun breakWord(text: CharSequence, beginIndex: Int, endIndex: Int): WordBreaker.WordBreaks {
        val length = endIndex - beginIndex
        val wordLevels = Reusables.wordLevels(length)

        if (length >= 3 && (patterns.count > 0 || exceptions.size > 0)) {
            val chars = Reusables.wordChars()
            if (!applyException(text, beginIndex, endIndex, wordLevels, chars)) {
                applyPatterns(text, beginIndex, endIndex, wordLevels, chars)
            }
        }

        return Reusables.wordBreaks(wordLevels, beginIndex)
    }

    private fun applyException(text: CharSequence, beginIndex: Int, endIndex: Int, wordLevels: PatternLevels, chars: PatternChars): Boolean {
        chars.reset(text, beginIndex, endIndex, true, true)
        val levels = exceptions[chars]
        if (levels != null) {
            levels.applyTo(wordLevels, 0)
            return true
        } else {
            return false
        }
    }

    private fun applyPatterns(text: CharSequence, beginIndex: Int, endIndex: Int, wordLevels: PatternLevels, chars: PatternChars) {
        for (begin in beginIndex..endIndex - 1) {
            for (end in begin + 1..min(endIndex, begin + patterns.maxLength())) {
                if (begin == beginIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, true, false)
                }
                if (end == endIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, false, true)
                }
                if (begin == beginIndex && end == endIndex) {
                    applyLevels(text, beginIndex, wordLevels, chars, begin, end, true, true)
                }
                applyLevels(text, beginIndex, wordLevels, chars, begin, end, false, false)
            }
        }
    }

    private fun applyLevels(text: CharSequence, beginIndex: Int, wordLevels: PatternLevels, chars: PatternChars, begin: Int, end: Int,
                            atWordBegin: Boolean, atWordEnd: Boolean) {
        chars.reset(text, begin, end, atWordBegin, atWordEnd)
        val levels = patterns.levelsFor(chars)
        levels?.applyTo(wordLevels, begin - beginIndex)
    }

    class Builder {
        private val patterns = Patterns()
        private val exceptions = CharsLevels()

        @Throws(IOException::class)
        fun addPatternsFrom(stream: InputStream): Builder {
            val patterns = readLines(InputStreamReader(stream, Charsets.UTF_8))
            for (pattern in patterns) {
                addPattern(pattern)
            }
            return this
        }

        @Throws(IOException::class)
        fun addExceptionsFrom(stream: InputStream): Builder {
            val patterns = readLines(InputStreamReader(stream, Charsets.UTF_8))
            for (pattern in patterns) {
                addException(pattern)
            }
            return this
        }

        fun addPattern(pattern: String): Builder {
            val letters = CharArrayList()
            val levels = PatternLevels()
            levels.resize(pattern.length + 1)

            var atWordBegin = false
            var atWordEnd = false

            for (i in 0..pattern.length - 1) {
                val ch = pattern[i]
                if (ch == EDGE_OF_WORD) {
                    if (i == 0) {
                        atWordBegin = true
                    } else if (i == pattern.length - 1) {
                        atWordEnd = true
                    }
                } else if ('0' <= ch && ch <= '9') {
                    levels.set(letters.size(), (ch - '0').toByte())
                } else {
                    letters.add(ch)
                }
            }

            val patternStr = String(letters.buffer, 0, letters.elementsCount)
            val patternChars = PatternChars(patternStr, 0, patternStr.length, atWordBegin, atWordEnd)

            levels.resize(letters.size() + 1)
            levels.trimToSize()

            patterns.put(patternChars, levels)

            return this
        }

        fun addException(exception: String): Builder {
            val letters = CharArrayList()
            val levels = PatternLevels()
            levels.resize(exception.length + 1)

            for (i in 0..exception.length - 1) {
                val ch = exception[i]
                if (ch == '-') {
                    levels.set(letters.size(), 9.toByte())
                } else {
                    letters.add(ch)
                }
            }

            val patternStr = String(letters.buffer, 0, letters.elementsCount)
            val patternChars = PatternChars(patternStr, 0, patternStr.length, true, true)

            levels.resize(letters.size() + 1)
            levels.trimToSize()

            exceptions.put(patternChars, levels)

            return this
        }

        fun build(): TeXHyphenator {
            return TeXHyphenator(patterns, exceptions)
        }
    }

    private class Patterns {
        private var lengthToCharsLevels = arrayOfNulls<CharsLevels>(16)
        private var maxLength = 0
        var count = 0

        fun put(chars: PatternChars, levels: PatternLevels) {
            val length = chars.length
            if (length > lengthToCharsLevels.size) {
                lengthToCharsLevels = copyOf<CharsLevels>(lengthToCharsLevels, length shr 1)
            }
            if (length > maxLength) {
                maxLength = length
            }

            var charsLevels: CharsLevels? = lengthToCharsLevels[length]
            if (charsLevels == null) {
                charsLevels = CharsLevels()
                lengthToCharsLevels[length] = charsLevels
            }

            charsLevels.put(chars, levels)
            count++
        }

        fun maxLength(): Int {
            return maxLength
        }

        fun levelsFor(chars: PatternChars): PatternLevels? {
            val charsLevels = lengthToCharsLevels[chars.length]
            return if (charsLevels != null) charsLevels[chars] else null
        }
    }

    private class CharsLevels : HashMap<PatternChars, PatternLevels>()

    @Reusable
    private class PatternChars {
        private var str: CharSequence? = null
        private var begin: Int = 0
        private var end: Int = 0
        private var atWordBegin: Boolean = false
        private var atWordEnd: Boolean = false

        var length: Int = 0
        private var hashCode: Int = 0

        constructor() {
        }

        constructor(str: CharSequence, begin: Int, end: Int, atWordBegin: Boolean, atWordEnd: Boolean) {
            reset(str, begin, end, atWordBegin, atWordEnd)
        }

        fun reset(str: CharSequence, begin: Int, end: Int, atWordBegin: Boolean, atWordEnd: Boolean) {
            this.str = str
            this.begin = begin
            this.end = end
            this.atWordBegin = atWordBegin
            this.atWordEnd = atWordEnd
            this.length = end - begin

            computeHashCode()
        }


        override fun equals(other: Any?): Boolean {
            other as PatternChars

            if (other.length != length || other.atWordBegin != atWordBegin || other.atWordEnd != atWordEnd) {
                return false
            }

            var i = begin
            var j = other.begin
            while (i < end) {
                if (str!![i] != other.str!![j]) {
                    return false
                }
                i++
                j++
            }

            return true
        }

        override fun hashCode(): Int {
            return hashCode
        }

        private fun computeHashCode() {
            hashCode = 0
            for (i in begin..end - 1) {
                hashCode = 31 * hashCode + str!![i].toInt()
            }
            hashCode = 31 * hashCode + if (atWordBegin) 1 else 0
            hashCode = 31 * hashCode + if (atWordEnd) 1 else 0
        }
    }

    @Reusable
    private class PatternLevels : ByteArrayList() {
        internal fun applyTo(levels: PatternLevels, beginIndex: Int) {
            var i = 0
            var j = beginIndex
            while (i < elementsCount && j < levels.elementsCount) {
                val level = get(i)
                if (level > levels.get(j)) {
                    levels.set(j, level)
                }
                i++
                j++
            }
        }
    }

    @Reusable
    private class WordBreaksImpl : WordBreaker.WordBreaks {
        private var levels: PatternLevels? = null
        private var beginIndex: Int = 0

        fun reset(levels: PatternLevels, beginIndex: Int) {
            this.levels = levels
            this.beginIndex = beginIndex
        }

        override fun canBreakBefore(index: Int): Boolean {
            val wordIndex = index - beginIndex
            // переносить одну букву нельзя
            val isMiddleBreak = wordIndex >= 2 && wordIndex < levels!!.elementsCount - 2
            return isMiddleBreak && levels!!.get(wordIndex) % 2 != 0
        }
    }

    private object Reusables {
        private val wordChars = reuser<PatternChars>({ PatternChars() })
        private val wordLevels = reuser<PatternLevels>({ PatternLevels() })
        private val wordBreaks = reuser<WordBreaksImpl>({ WordBreaksImpl() })

        fun wordChars(): PatternChars {
            return wordChars.reuse()
        }

        fun wordLevels(wordLength: Int): PatternLevels {
            val value = wordLevels.reuse()
            value.clear()
            value.resize(wordLength + 1)
            return value
        }

        fun wordBreaks(levels: PatternLevels, beginIndex: Int): WordBreaker.WordBreaks {
            val value = wordBreaks.reuse()
            value.reset(levels, beginIndex)
            return value
        }
    }

    companion object {
        /**
         * алгоритм: https://habrahabr.ru/post/138088/
         */

        private val EDGE_OF_WORD = '.'
    }
}
